hudi-client/hudi-client-common/src/main/java/org/apache/hudi/optimize/HilbertCurve.java

```diff
@@ -0,0 +1,321 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.optimize;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+
+/**
+ * Converts between Hilbert index ({@code BigInteger}) and N-dimensional points.
+ *
+ * <p>
+ * Note:
+ * <a href="https://github.com/davidmoten/hilbert-curve/blob/master/src/main/java/org/davidmoten/hilbert/HilbertCurve.java">GitHub</a>).
+ * the Licensed of above link is also http://www.apache.org/licenses/LICENSE-2.0
+ */
+public final class HilbertCurve {
+
+  private final int bits;
+  private final int dimensions;
+  // cached calculations
+  private final int length;
+
+  private HilbertCurve(int bits, int dimensions) {
+    this.bits = bits;
+    this.dimensions = dimensions;
+    // cache a calculated values for small perf improvements
+    this.length = bits * dimensions;
+  }
+
+  /**
+   * Returns a builder for and object that performs transformations for a Hilbert
+   * curve with the given number of bits.
+   *
+   * @param bits
+   *            depth of the Hilbert curve. If bits is one, this is the top-level
+   *            Hilbert curve
+   * @return builder for object to do transformations with the Hilbert Curve
+   */
+  public static Builder bits(int bits) {
+    return new Builder(bits);
+  }
+
+  /**
+   * Builds a {@link HilbertCurve} instance.
+   */
+  public static final class Builder {
+    final int bits;
+
+    private Builder(int bits) {
+      if (bits <= 0  || bits >= 64) {
+        throw new IllegalArgumentException(String.format("bits must be greater than zero and less than 64, now found bits value: %s", bits));
+      }
+      this.bits = bits;
+    }
+
+    public HilbertCurve dimensions(int dimensions) {
+      if (dimensions < 2) {
+        throw new IllegalArgumentException(String.format("dimensions must be at least 2, now found dimensions value: %s", dimensions));
+      }
+      return new HilbertCurve(bits, dimensions);
+    }
+  }
+
+  /**
+   * Converts a point to its Hilbert curve index.
+   *
+   * @param point
+   *            an array of {@code long}. Each ordinate can be between 0 and
+   *            2<sup>bits</sup>-1.
+   * @return index (nonnegative {@link BigInteger})
+   * @throws IllegalArgumentException
+   *             if length of point array is not equal to the number of
+   *             dimensions.
+   */
+  public BigInteger index(long... point) {
```

Comment:
Let's pass an array instead of var-args (var-arg is a better fit for loose-form method args, rather than well-scoped N-dimensional point)

Reply:
agree